### PR TITLE
#2024 Constraint changes

### LINKED ISSUE #2024: CUSTOMER - Constraints on OEntities and OActors should be displayed in referencing elements inside the Semantic Browser view
Constraints applied on the selected element should be displayed in the Semantic Browser (left-hand part: Constraining Elements).
They are not displayed on Operational Entities. Same problem for Operational Actors (Operational Entity [Actor]).

Note that they are correctly displayed on System Actors, the System, Operational Activities and Interactions.

### PR DESCRIPTION
**Pull request by claudiasima-ro**
Fixes #2024 

Change-Id: Ia7e224dc5f32fd26cdbd7112c64a78db3e120486
Signed-off-by: claudiasima <claudia.sima@thalesgroup.com>

### CODE REVIEW

File: core/plugins/org.polarsys.capella.core.data.core.validation/plugin.xml [hunk 1]
```diff
@@ -117,6 +117,44 @@
...
+               <message>
+                  {0} should be located under {1}.
+               </message>
+               <target
+                     class="Constraint">
+               </target>
+               <description>
+                  This rule ensures that if a constraint is applied to only one element then the constraint is located under Component.
+               </description>
+            </constraint>
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.core.validation/plugin.xml:157]: don't forget to regenerate validation rules documentation

File: core/plugins/org.polarsys.capella.core.data.core.validation/src/org/polarsys/capella/core/data/capellamodeller/validation/ConstraintLocationRule.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    EMFEventType eType = ctx.getEventType();
+    if (eType == EMFEventType.NULL && eObj instanceof Constraint) {
+      Constraint constraint = (Constraint) eObj;
+      EList<ModelElement> constrainedElements = constraint.getConstrainedElements();
+      
+      boolean isConstrainingParts = constrainedElements.stream().filter(element -> element instanceof Part)
+          .collect(Collectors.toList()).size() > 0;
+      boolean isConstrainingDeployments = constrainedElements.stream().filter(element -> element instanceof PartDeploymentLink)
+          .collect(Collectors.toList()).size() > 0;
+      if (constraint.eContainer() instanceof Part && !isConstrainingParts && !isConstrainingDeployments ||
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.core.validation/src/org/polarsys/capella/core/data/capellamodeller/validation/ConstraintLocationRule.java:47]: to ease readibility and prevent unexepected boolean &|& issues, can you add surrounding parenthesis

```((constraint.eContainer() instanceof Part && !isConstrainingParts && !isConstrainingDeployments) || (constraint.eContainer() instanceof PartDeploymentLink && !isConstrainingDeployments))```

**Review by pdulth — CHANGES_REQUESTED** [review]

File: core/plugins/org.polarsys.capella.core.data.core.ui.quickfix/src/org/polarsys/capella/core/data/core/ui/quickfix/resolver/ConstraintLocationRule_Resolver.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    Constraint constraint = (Constraint) tgts.get(0);
+    ModelElement comp = getConstraintNewLocation(constraint);
+
+    // move constraint to new location and delete it from current location
+    if (comp != null) {
+      AbstractReadWriteCommand abstractCommand = new AbstractReadWriteCommand() {
+        @Override
+        public void run() {
+          EObject currentConstraintLocation = getConstraintLocation(constraint);
+          if (currentConstraintLocation instanceof Part) {
```
[the remark this answers is not in the document]
Comment (claudiasima-ro) [reply core/plugins/org.polarsys.capella.core.data.core.ui.quickfix/src/org/polarsys/capella/core/data/core/ui/quickfix/resolver/ConstraintLocationRule_Resolver.java]: Done.

File: core/plugins/org.polarsys.capella.core.data.core.validation/src/org/polarsys/capella/core/data/capellamodeller/validation/ConstraintLocationRule.java (hunk 2 above)
Comment (claudiasima-ro) [reply core/plugins/org.polarsys.capella.core.data.core.validation/src/org/polarsys/capella/core/data/capellamodeller/validation/ConstraintLocationRule.java]: Done.

**Review by georgiana-ecobici — APPROVED** [review]

**Review by pdulth — APPROVED** [review]

**Pull request merged**